homeassistant/components/media_player/pjlink.py

```diff
@@ -0,0 +1,156 @@
+"""
+import logging
+
+import voluptuous as vol
+
+from homeassistant.components.media_player import (
+    SUPPORT_TURN_OFF, SUPPORT_TURN_ON, SUPPORT_VOLUME_MUTE,
+    SUPPORT_SELECT_SOURCE, PLATFORM_SCHEMA, MediaPlayerDevice)
+from homeassistant.const import (
+    STATE_OFF, STATE_ON, CONF_HOST,
+    CONF_NAME, CONF_PASSWORD, CONF_PORT)
+import homeassistant.helpers.config_validation as cv
+
+REQUIREMENTS = ['pypjlink2==1.2.0']
+
+_LOGGER = logging.getLogger(__name__)
+
+CONF_ENCODING = 'encoding'
+
+DEFAULT_PORT = 4352
+DEFAULT_ENCODING = 'utf-8'
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_HOST): cv.string,
+    vol.Optional(CONF_PORT, default=DEFAULT_PORT): cv.port,
+    vol.Optional(CONF_NAME): cv.string,
+    vol.Optional(CONF_ENCODING, default=DEFAULT_ENCODING): cv.string,
+    vol.Optional(CONF_PASSWORD): cv.string,
+})
+
+SUPPORT_PJLINK = SUPPORT_VOLUME_MUTE | \
+    SUPPORT_TURN_ON | SUPPORT_TURN_OFF | SUPPORT_SELECT_SOURCE
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the PJLink platform."""
+    host = config.get(CONF_HOST)
+    port = config.get(CONF_PORT)
+    name = config.get(CONF_NAME)
+    encoding = config.get(CONF_ENCODING)
+    password = config.get(CONF_PASSWORD)
+
+    if 'pjlink' not in hass.data:
+        hass.data['pjlink'] = {}
+    hass_data = hass.data['pjlink']
+
+    device_label = "{}:{}".format(host, port)
+    if device_label in hass_data:
+        return
+
+    device = PjLinkDevice(host, port, name, encoding, password)
+    hass_data[device_label] = device
+    add_devices([device], True)
+
+
+class PjLinkDevice(MediaPlayerDevice):
+    """Representation of a PJLink device."""
+
+    def __init__(self, host, port, name, encoding, password):
+        """Iinitialize the PJLink device."""
+        self._host = host
+        self._port = port
+        self._name = name
+        self._password = password
+        self._encoding = encoding
+        self._muted = False
+        self._pwstate = STATE_OFF
+        self._current_source = None
+        with self.projector() as projector:
+            if not self._name:
+                self._name = projector.get_name()
+            inputs = projector.get_inputs()
+        self._source_name_mapping = \
+            {self.format_input_source(*x): x for x in inputs}
+        self._source_list = sorted(self._source_name_mapping.keys())
+
+    def projector(self):
+        """Create PJLink Projector instance."""
+        from pypjlink import Projector
+        projector = Projector.from_address(self._host, self._port,
+                                           self._encoding)
+        projector.authenticate(self._password)
+        return projector
+
+    def format_input_source(self, input_source_name, input_source_number):
+        """Format input source for display in UI"""
+        return "{} {}".format(input_source_name, input_source_number).title()
```

Comment:
Please format this as all lowercase snakecase. The backend doesn't care about frontend formatting. We need consistency in string formats for users in templates etc and for possible future translations.

Reply:
I can convert to snake case but how to I make the UI display something like `Digital 1` instead of `digital_1`?

Reply:
Also the source name are returned programmatically from the device and could be different from a projector to another depending on the model or if the manufacturer chooses to name the inputs differently. A projector could also provides the option to customize the name of the inputs.

I was looking at the sonos component where the input sources are the favorites defined by the user in the sonos app. For that component the source names are not snake case. In my opinion this is a similar case and we should just return the name of the source as returned by the projector.

Reply:
I don't know so much about the frontend or translation project. So it's best to ask someone else eg in frontend dev chat. For static strings, something like this is probably what it takes:
https://github.com/home-assistant/home-assistant/pull/15498

Reply:
Backend strings should be lowercase snakecase, period. Older bad examples doesn't matter.

Reply:
@MartinHjelmare backend strings are now lower snake case. I've also asked about dynamic backend strings and I don't think that this is something that has been thought through.
I opened this issue in the architecture project https://github.com/home-assistant/architecture/issues/47